File: backend-csharp/Controllers/JobsController.cs
using Microsoft.AspNetCore.Mvc;
using MongoDB.Bson;
using MongoDB.Driver;

namespace backend_csharp.Controllers
{
    [ApiController]
    [Route("api/v1/jobs")]
    public class JobsController : ControllerBase
    {
        private readonly IMongoCollection<BsonDocument> _jobs;

        public JobsController()
        {
            var connectionString = Environment.GetEnvironmentVariable("MongoDbConnectionString");
            var databaseName = Environment.GetEnvironmentVariable("MongoDbDatabaseName");
            var client = new MongoClient(connectionString);
            var db = client.GetDatabase(databaseName);
            _jobs = db.GetCollection<BsonDocument>("JobApplications");
        }

        // ✅ GET /api/v1/jobs?userDataID=IP083
        [HttpGet]
        public IActionResult GetAllJobsForUser([FromQuery] string userDataID)
        {
            var filter = Builders<BsonDocument>.Filter.Eq("userDataID", userDataID);
            var jobs = _jobs.Find(filter).ToList();

            // Convert BsonDocument to dictionary to avoid serialization errors
            var jobList = jobs.Select(job => job.ToDictionary()).ToList();

            return Ok(jobList);
        }


        //curl -Method POST "http://localhost:5135/api/v1/jobs" `
        //-Headers @{ "Content-Type" = "application/json" } `
        //-Body '{"jobId":"IP083-3163","jobTitle":"Junior SWE","company":"CIBC","status":"Applied","dateApplied":"2025-09-10","accountNeeded":false,"jobUrl":"https://www.linkedin.com/jobs/view/4298287896","userDataID":"IP083"}'

        [HttpPost]
        public IActionResult AddJob([FromBody] JobRequest request)
        {
            var jobDoc = new BsonDocument {
                { "jobId", request.JobID },
                { "jobTitle", request.JobTitle },
                { "company", request.Company },
                { "status", request.Status },
                { "dateApplied", request.DateApplied },
                { "accountNeeded", request.AccountNeeded },
                { "jobUrl", request.JobUrl },
                { "userDataID", request.UserDataID }
            };

            _jobs.InsertOne(jobDoc);
            return Ok(new { jobId = request.JobID });

        }

        [HttpPut("{id}")]
        public IActionResult UpdateJob(string id, [FromBody] JobRequest request)
        {
            var filter = Builders<BsonDocument>.Filter.Eq("_id", ObjectId.Parse(id)) &
                         Builders<BsonDocument>.Filter.Eq("userDataID", request.UserDataID);

            var update = Builders<BsonDocument>.Update
                .Set("jobTitle", request.JobTitle)
                .Set("company", request.Company)
                .Set("status", request.Status)
                .Set("dateApplied", request.DateApplied)
                .Set("accountNeeded", request.AccountNeeded)
                .Set("jobUrl", request.JobUrl);

            var result = _jobs.UpdateOne(filter, update);
            if (result.MatchedCount == 0) return NotFound();

            return Ok(new { updated = true });
        }

        [HttpDelete("{jobId}")]
        public IActionResult DeleteJob(string jobId, [FromQuery] string userDataID)
        {
            var filter = Builders<BsonDocument>.Filter.Eq("jobId", jobId) &
                        Builders<BsonDocument>.Filter.Eq("userDataID", userDataID);

            var result = _jobs.DeleteOne(filter);
            if (result.DeletedCount == 0) return NotFound();

            return Ok(new { deleted = true });
        }

    }

    public class JobRequest
    {
        public string JobID { get; set; }
        public string JobTitle { get; set; }
        public string Company { get; set; }
        public string Status { get; set; }
        public string DateApplied { get; set; }
        public bool AccountNeeded { get; set; }
        public string JobUrl { get; set; }
        public string UserDataID { get; set; }
    }
}

File: backend-springboot/src/main/java/com/IshanPhadte/ApplyToJobDashboard/controller/JobsController.java
package com.IshanPhadte.ApplyToJobDashboard.controller;

import java.util.List;
import java.util.Optional;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.IshanPhadte.ApplyToJobDashboard.model.JobApplication;
import com.IshanPhadte.ApplyToJobDashboard.repository.JobApplicationRepository;

@RestController
@RequestMapping("/api/v1/jobs")
public class JobsController {

    private final JobApplicationRepository repository;

    public JobsController(JobApplicationRepository repository) {
        this.repository = repository;
    }

    /*
    Sample PowerShell curl commands for testing:

    # 1️⃣ GET all jobs for userDataID IP083
    curl -Method GET "http://localhost:8080/api/v1/jobs?userDataID=IP083" `
         -Headers @{ "Content-Type" = "application/json" }

    # 2️⃣ POST add a new job
    curl -Method POST "http://localhost:8080/api/v1/jobs" `
         -Headers @{ "Content-Type" = "application/json" } `
         -Body '{ 
             "jobId":"IP083-3163",
             "jobTitle":"Junior SWE",
             "company":"CIBC",
             "status":"Applied",
             "dateApplied":"2025-09-10",
             "accountNeeded": false,
             "jobUrl":"https://www.linkedin.com/jobs/view/4298287896",
             "userDataID":"IP083"
         }'

    # 3️⃣ PUT update a job by MongoDB _id
    curl -Method PUT "http://localhost:8080/api/v1/jobs/{id}" `
         -Headers @{ "Content-Type" = "application/json" } `
         -Body '{ 
             "jobId":"IP083-3163",
             "jobTitle":"Junior SWE Updated",
             "company":"CIBC",
             "status":"Interview Scheduled",
             "dateApplied":"2025-09-10",
             "accountNeeded": true,
             "jobUrl":"https://www.linkedin.com/jobs/view/4298287896",
             "userDataID":"IP083"
         }'

    # 4️⃣ DELETE job by jobId and userDataID
    curl -Method DELETE "http://localhost:8080/api/v1/jobs/IP083-3163?userDataID=IP083" `
         -Headers @{ "Content-Type" = "application/json" }
    */

    // GET all jobs for a user
    @GetMapping
    public List<JobApplication> getAllJobsForUser(@RequestParam String userDataID) {
        return repository.findByUserDataID(userDataID);
    }

    // POST add a new job
    @PostMapping
    public JobApplication addJob(@RequestBody JobApplication request) {
        return repository.save(request);
    }

    // PUT update job by MongoDB _id
    // PUT update job by jobId
    @PutMapping("/update/{jobId}")
    public ResponseEntity<?> updateJobByJobId(@PathVariable String jobId, @RequestBody JobApplication request) {
        Optional<JobApplication> existing = repository.findByJobIdAndUserDataID(jobId, request.getUserDataID());
        if (existing.isEmpty()) {
            return ResponseEntity.notFound().build();
        }

        JobApplication jobToUpdate = existing.get();
        jobToUpdate.setJobTitle(request.getJobTitle());
        jobToUpdate.setCompany(request.getCompany());
        jobToUpdate.setStatus(request.getStatus());
        jobToUpdate.setDateApplied(request.getDateApplied());
        jobToUpdate.setAccountNeeded(request.isAccountNeeded());
        jobToUpdate.setJobUrl(request.getJobUrl());

        repository.save(jobToUpdate);
        return ResponseEntity.ok("{\"updated\": true}");
    }


    // DELETE job by jobId and userDataID
    @DeleteMapping("/{jobId}")
    public ResponseEntity<?> deleteJob(@PathVariable String jobId, @RequestParam String userDataID) {
        Optional<JobApplication> existing = repository.findByJobIdAndUserDataID(jobId, userDataID);
        if (existing.isEmpty()) return ResponseEntity.notFound().build();

        repository.delete(existing.get());
        return ResponseEntity.ok("{\"deleted\": true}");
    }
}
